src/main/java/com/salesforce/mirus/HerderStatusMonitor.java

```diff
@@ -91,42 +90,52 @@ private void onConnectors(Throwable error, Collection<String> connectorNames) {
-          logger.info("Attempting to restart connector {}", connectorName);
-          herder.restartConnector(
-              connectorName,
-              (error, _void) -> {
-                if (error != null) {
-                  logger.warn("Failed to restart connector {}", connectorName, error);
-                }
-              });
-        }
-      } else {
-        herder.connectorInfo(connectorName, this::onConnectInfo);
-      }
-    }
-  }
-
-  private void onConnectInfo(Throwable error, ConnectorInfo connectorInfo) {
-    if (error != null) {
-      logger.warn("Failed to retrieve connector info, Error details: {}", error);
-      return;
+    boolean isAssignedWorker = workerId.equals(stateInfo.connector().workerId());
+    State connectorState = State.valueOf(stateInfo.connector().state());
+
+    herder.connectorInfo(
+        connectorName,
+        (error, connectorInfo) -> {
+          if (error != null) {
+            logger.warn("Failed to retrieve connector info, Error details: {}", error);
+            return;
+          }
+          // Only the worker assigned to this controller should report connector metrics
+          // Report metrics regardless of current state
+          if (isAssignedWorker) {
+            connectorJmxReport.handleConnector(herder, connectorInfo);
+          }
+
+          if (connectorState == State.RUNNING) {
+            // All workers need to process the all assigned tasks for the current connector
+            connectorInfo.tasks().forEach(task -> processTask(task, herder.taskStatus(task)));
+          }
+        });
+
+    // Only the assigned worker should attempt to restart a failed connector
+    if (autoRestartConnectorEnabled && isAssignedWorker && connectorState == State.FAILED) {
+      logger.info("Attempting to restart connector {}", connectorName);
+      connectorJmxReport.incrementConnectorRestartAttempts(connectorName);
+      herder.restartConnector(
+          connectorName,
+          (error, _void) -> {
+            if (error != null) {
+              logger.warn("Failed to restart connector {}", connectorName, error);
+            }
+          });
     }
-
-    connectorJmxReport.handleConnector(herder, connectorInfo);
-    connectorInfo.tasks().forEach(task -> processTask(task, herder.taskStatus(task)));
   }
 
   private void processTask(ConnectorTaskId taskId, TaskState taskStatus) {
-    taskJmxReporter.updateMetrics(taskId, taskStatus);
-    if (taskStatus.state().equalsIgnoreCase(TaskStatus.State.FAILED.toString())) {
-      connectorJmxReport.incrementTotalFailedCount(taskId.connector());
-      if (autoRestartTaskEnabled) {
-        logger.info("Attempting to restart task {}", taskId);
-        herder.restartTask(taskId, this::onTaskRestart);
+
+    if (workerId.equals(taskStatus.workerId())) {
+      taskJmxReporter.updateMetrics(taskId, taskStatus);
+      if (taskStatus.state().equalsIgnoreCase(TaskStatus.State.FAILED.toString())) {
```

Comment:
Can you also update this to use the AbstractStatus.State?

Reply:
Yes.  I've updated to explicitly use Task and Connector status too.